### PR TITLE
feat: add MTP (multi-token prediction) training support for nemotron

### PR DESCRIPTION
**Pull request by yfw**
# What does this PR do ?

Adds training-side Megatron Multi-Token Prediction support, rebased onto the Megatron-Bridge 79c8957 bump which includes needed Megatron-LM changes. Part of Nemotron 3 Super upstreaming.

- megatron_policy_worker: compute mtp_loss_mask from token_mask × sample_mask (with optional mtp_positive_only gating on advantages), set mtp_grad_scale_func for MTP loss scaling, and collect MTP metrics.
- data.py / train.py: plumb mtp_loss_mask through microbatch processing/packing into model_forward's loss_mask.
- common.py: get_mtp_metrics (per-layer MTP loss + acceptance rate).
- grpo.py / lm_policy.py: surface mtp/* metrics in training logs.

Scope: training only. The MTP inference path (vLLM drafter weight load / speculative-config handling) will be handled separately as part of the ultra upstreaming.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/models/megatron/common.py [hunk 1]
```diff
@@ -164,3 +165,34 @@ def get_moe_metrics(
...
+        mtp_losses = tracker["loss_values"].float()
+        mtp_corrects = tracker.get("correct_values", torch.zeros_like(mtp_losses))
+        mtp_totals = tracker.get("total_values", torch.ones_like(mtp_losses))
+        mtp_num_layers = mtp_losses.shape[0]
+
+        # Log per-layer losses and acceptance rates
+        for i in range(mtp_num_layers):
+            metrics[f"mtp_{i + 1}_loss"] = float(mtp_losses[i].item())
+            # Compute acceptance rate as percentage
+            acceptance_rate = (mtp_corrects[i] / max(mtp_totals[i], 1)) * 100.0
```
Comment (ashors1) [inline nemo_rl/models/megatron/common.py]: [`common.py:194`](https://github.com/NVIDIA-NeMo/RL/blob/f22fecea6e6012e542edf1f900c80988dc108d46/nemo_rl/models/megatron/common.py#L194)

`max(mtp_totals[i], 1)` uses Python's built-in `max()` with a 0-dim CUDA tensor, producing inconsistent return types (Python `int` vs tensor) depending on which is larger. The upstream Megatron-LM code uses `torch.clamp()` for this pattern.

```suggestion
            acceptance_rate = (mtp_corrects[i] / mtp_totals[i].clamp(min=1)) * 100.0
```

File: nemo_rl/models/megatron/data.py [hunk 2]
```diff
@@ -302,6 +306,24 @@ def process_microbatch(
                     cp_size=get_context_parallel_world_size(),
                 )
 
+                # Pack pre-computed mtp_loss_mask the same way as input_ids
```
Comment (ashors1) [inline nemo_rl/models/megatron/data.py:390]: [`data.py:284`](https://github.com/NVIDIA-NeMo/RL/blob/f22fecea6e6012e542edf1f900c80988dc108d46/nemo_rl/models/megatron/data.py#L284)

The VLM branch (`_prepare_vlm_batch_for_megatron`, a few lines above) doesn't pack or propagate `mtp_loss_mask`. If someone uses VLM + sequence packing + MTP, the loss mask is silently dropped with no error. Consider adding a guard before that branch:

```python
if self._has_vlm_data(data_dict):
    assert "mtp_loss_mask" not in data_dict, (
        "MTP training is not supported with VLM sequence packing"
    )
```

File: nemo_rl/models/megatron/data.py [hunk 3]
```diff
@@ -66,6 +67,7 @@ class ProcessedMicrobatch:
     position_ids: Optional[torch.Tensor]
     packed_seq_params: Optional[PackedSeqParams]
     cu_seqlens_padded: Optional[torch.Tensor]
+    mtp_loss_mask: Optional[torch.Tensor] = None
```
Comment (ashors1) [inline nemo_rl/models/megatron/data.py:80]: [`data.py:61`](https://github.com/NVIDIA-NeMo/RL/blob/f22fecea6e6012e542edf1f900c80988dc108d46/nemo_rl/models/megatron/data.py#L61)

`ProcessedMicrobatch` has an `Attributes:` docstring section that lists all fields, but `mtp_loss_mask` is missing from it. Could you add it after `cu_seqlens_padded`:

```
        mtp_loss_mask: Pre-computed MTP loss mask (token_mask × sample_mask).
            None when MTP is disabled or token/sample masks are absent.
```

File: tests/unit/models/megatron/test_megatron_data.py [hunk 4]
```diff
@@ -232,6 +233,88 @@ def test_process_microbatch_with_packing(
...
+            seq_length_key="input_lengths",
+            pack_sequences=True,
+            straggler_timer=MagicMock(),
+        )
+
+        # _pack_sequences_for_megatron is called once for input_ids and once for mtp_loss_mask.
+        assert mock_pack.call_count == 2
+        # mtp_loss_mask takes the packed tensor (index 1 of the pack return tuple).
+        assert result.mtp_loss_mask is not None
+        assert torch.equal(result.mtp_loss_mask, packed)
```
Comment (ashors1) [inline tests/unit/models/megatron/test_megatron_data.py]: [`test_megatron_data.py:316`](https://github.com/NVIDIA-NeMo/RL/blob/f22fecea6e6012e542edf1f900c80988dc108d46/tests/unit/models/megatron/test_megatron_data.py#L316)

The mock returns the same `packed` tensor at both index 0 and index 1, so this assertion passes regardless of whether the implementation reads index 0 or index 1 — it can’t catch an off-by-one index bug. Consider using two distinct tensors:

```python
packed_idx0 = torch.zeros(1, 8, dtype=torch.long)
packed_idx1 = torch.tensor([[1, 2, 3, 4, 5, 6, 7, 8]])
mock_pack.return_value = (
    packed_idx0, packed_idx1, MagicMock(),
    torch.tensor([0, 5, 8], dtype=torch.int32),
    torch.tensor([0, 5, 8], dtype=torch.int32),
)
# ...
assert torch.equal(result.mtp_loss_mask, packed_idx1)  # confirms index 1
```

File: nemo_rl/models/megatron/common.py [hunk 5]
```diff
@@ -21,6 +21,7 @@
     get_moe_layer_wise_logging_tracker,
     reduce_aux_losses_tracker_across_ranks,
 )
+from megatron.core.transformer.multi_token_prediction import MTPLossLoggingHelper
```
Comment (ashors1) [inline nemo_rl/models/megatron/common.py:24]: [`common.py:1`](https://github.com/NVIDIA-NeMo/RL/blob/f22fecea6e6012e542edf1f900c80988dc108d46/nemo_rl/models/megatron/common.py#L1)

This file (and `data.py`, `megatron_policy_worker.py`, `__init__.py`, `lm_policy.py`, `grpo.py`) still has `# Copyright (c) 2025`. Current year is 2026 — `train.py` and `setup.py` in this same PR were already updated.

**Review by ashors1 — COMMENTED** [review]
Thanks for the MTP training support — clean implementation with good test coverage!

A few findings below (5 inline comments). The VLM guard and test assertion are the most actionable.

**Copyright note (spans multiple files):** `common.py`, `data.py`, `megatron_policy_worker.py`, `__init__.py`, `lm_policy.py`, and `grpo.py` still carry `# Copyright (c) 2025`. Current year is 2026 — `train.py` and `setup.py` were already correctly updated in this PR.

_Generated by Claude Code_

File: nemo_rl/models/policy/workers/megatron_policy_worker.py [hunk 6]
```diff
@@ -1207,6 +1226,36 @@ def prepare_refit_info(self) -> None:
...
+
+        get_mtp_metrics is imported lazily (not a module global) so cloudpickle
+        does not pull an unpicklable torch ConfigModuleInstance into the worker
+        actor's serialization.
+        """
+        mtp_num_layers = getattr(self.model.config, "mtp_num_layers", None)
+        if mtp_num_layers is not None and mtp_num_layers > 0:
+            from nemo_rl.models.megatron.common import get_mtp_metrics
+
+            mtp_metrics = get_mtp_metrics()
```
Comment (macandro96) [inline nemo_rl/models/policy/workers/megatron_policy_worker.py:1307]: Do we need to broadcast mtp metrics from to other stages if `pipeline_model_parallel_size > 1`? Similar to [here](https://github.com/NVIDIA-NeMo/RL/blob/f22fecea6e6012e542edf1f900c80988dc108d46/nemo_rl/models/policy/workers/megatron_policy_worker.py#L654)

File: nemo_rl/models/megatron/data.py (hunk 2 above)
Comment (yfw) [reply nemo_rl/models/megatron/data.py:390]: Addressed in 0981ab0119a9eae6f0a3643e6969fdfb28f2f5b8

File: tests/unit/models/megatron/test_megatron_data.py (hunk 4 above)
Comment (yfw) [reply tests/unit/models/megatron/test_megatron_data.py]: Addressed in 0981ab0119a9eae6f0a3643e6969fdfb28f2f5b8

File: nemo_rl/models/megatron/common.py (hunk 5 above)
Comment (yfw) [reply nemo_rl/models/megatron/common.py:24]: Addressed in 0981ab0119a9eae6f0a3643e6969fdfb28f2f5b8

File: nemo_rl/models/policy/workers/megatron_policy_worker.py (hunk 6 above)
Comment (yfw) [reply nemo_rl/models/policy/workers/megatron_policy_worker.py:1307]: Addressed in 0981ab0119a9eae6f0a3643e6969fdfb28f2f5b8

**Review by macandro96 — APPROVED** [review]

File: nemo_rl/models/megatron/data.py (hunk 3 above)
Comment (yfw) [reply nemo_rl/models/megatron/data.py:80]: Addressed in https://github.com/NVIDIA-NeMo/RL/commit/0981ab0119a9eae6f0a3643e6969fdfb28f2f5b8

**Review by ashors1 — APPROVED** [review]
